File: CCO/CCO '15 P2 - Artskjid.py
# 49/50 using bitmask DP
# Even though the time complexity is exponential, there are still overlapping states so we can use DP
# Instead of storing each state as a list of True/False, we can use a number (0 and 1 in binary) to
# represent if a node has been visited

n_nodes, n_roads = map(int, input().split())
graph = [[] for _ in range(n_nodes)]

for _ in range(n_roads):
    a, b, dist = map(int, input().split())
    graph[a].append((b, dist))

res = 0
max_dists = [[0] * (2 ** 18) for _ in range(n_nodes)]

stack = [(0, 1, 0)]  # current node, path dp state, distance
while stack:
    node, path, dist = stack.pop()
    if node == n_nodes - 1:  # new way to get to end, update answer
        res = max(res, dist)
        continue

    for adj, adj_dist in graph[node]:
        new_dist = dist + adj_dist
        if path & (1 << adj):  # already visited this node on the current path
            continue

        new_path = path | (1 << adj)  # mark as visited in current path

        if new_dist > max_dists[adj][new_path]:  # found better path
            max_dists[adj][new_path] = new_dist
            stack.append((adj, new_path, new_dist))

print(res)

# # 43/50 test cases, TLE
# # store each state as (current node, visited nodes)
#
# n_nodes, n_roads = map(int, input().split())
# graph = [[] for _ in range(n_nodes)]
#
# for _ in range(n_roads):
#     a, b, dist = map(int, input().split())
#     graph[a].append((b, dist))
#
# max_dist = 0
# state = [False for _ in range(n_nodes)]
# state[0] = True
#
# memo = {}  # keep track of already visited paths
# q = [(0, tuple(state), 0)]
# while q:
#     node, path, dist = q.pop()
#     if node == n_nodes - 1:
#         max_dist = max(max_dist, dist)
#         continue
#
#     for adj, adj_dist in graph[node]:
#         new_dist = dist + adj_dist
#         if path[adj]:
#             continue
#
#         new_path = list(path)
#         new_path[adj] = True
#         new_path = tuple(new_path)
#
#         if (adj, new_path) not in memo or new_dist > memo[(adj, new_path)]:
#             memo[(adj, new_path)] = new_dist
#             q.append((adj, new_path, new_dist))
#
# print(max_dist)

File: CCO/CCO '15 P2 - Artskjid.cpp
// https://dmoj.ca/problem/cco15p2
// bitmask DP with DFS
// store the nodes we visited on the current path as a bitmask
// dp[i][mask]: longest path from the i-th node, already visited set bits in mask

#include <bits/stdc++.h>

#define pii pair<int, int>

using namespace std;

const int inf = 1000000000;
int N, M;
vector<vector<pii>> graph;
vector<vector<int>> dist;

int solve(int cur, int mask) {
    if (cur == N - 1)  // reached destination
        return 0;
    if (dist[cur][mask] != -1)  // cache
        return dist[cur][mask];
    int mx = -inf;
    for (auto [adj, adj_d]: graph[cur]) {
        if (mask & (1 << adj))  // already visited
            continue;
        mx = max(mx, adj_d + solve(adj, mask | (1 << adj)));
    }
    return dist[cur][mask] = mx;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    graph.resize(N, vector<pii >());
    dist.resize(N, vector<int>(1 << N, -1));

    for (int i = 0; i < M; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        graph[a].emplace_back(b, c);
    }

    cout << solve(0, 1) << "\n";
    return 0;
}
